File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def findPath(self, root: TreeNode, node: TreeNode, path: List[TreeNode]) -> List[TreeNode]:
        path.append(root)

        if node.val == root.val:
            return path
        elif node.val > root.val:
            return self.findPath(root.right, node, path)
        else:
            return self.findPath(root.left, node, path)

    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        p_path = self.findPath(root, p, [])
        q_path = self.findPath(root, q, [])

        result = None

        for i in range(0, min(len(p_path), len(q_path))):
            if p_path[i] != q_path[i]:
                result = p_path[i - 1]
                break

        if not result:
            result = p_path[-1] if len(p_path) < len(q_path) else q_path[-1]
       
        return result

File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
   public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        if (!root) return {};

        deque<TreeNode*> q{};
        q.push_back(root);
        vector<int> layer{};
        vector<vector<int>> result{};

        int level_size = q.size();

        while (!q.empty()) {
            for (int i = 0; i < level_size; ++i) {
                TreeNode* node = q.front();
                q.pop_front();

                layer.push_back(node->val);

                if (node->left) {
                    q.push_back(node->left);
                }
                if (node->right) {
                    q.push_back(node->right);
                }
            }

            result.push_back(layer);
            layer.clear();
            level_size = q.size();
        }

        return result;
    }
};
